From 125f5531d420005cbc1198df4a7daef6bd9e42d6 Mon Sep 17 00:00:00 2001
From: Sharaf Zaman <sharafzaz121@gmail.com>
Date: Tue, 11 Feb 2020 18:12:20 +0300
Subject: [PATCH 2/3] Bugfix: fix the offset bug when using Stylus with Android

Platform plugin for Android calculated the local coordinates of the
tabletEvent by using QWindow::position(), which returned High-DPI
scaled coordinates. But, globalPosF being unscaled made the
calculation invalid.

Proper demonstration of this can be seen in Krita Android.

[ChangeLog][Android] Fix offset bug in tabletEvents

Change-Id: I4679df7de3b1491009708d7e80d5bc176e29afed
---
 src/plugins/platforms/android/androidjniinput.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/android/androidjniinput.cpp b/src/plugins/platforms/android/androidjniinput.cpp
index 9018d5b034..c57ff9ad59 100644
--- a/src/plugins/platforms/android/androidjniinput.cpp
+++ b/src/plugins/platforms/android/androidjniinput.cpp
@@ -43,6 +43,7 @@
 #include "androidjniinput.h"
 #include "androidjnimain.h"
 #include "qandroidplatformintegration.h"
+#include "qandroidplatformwindow.h"
 
 #include <qpa/qwindowsysteminterface.h>
 #include <QTouchEvent>
@@ -310,7 +311,12 @@ namespace QtAndroidInput
         QPointF globalPosF(x, y);
         QPoint globalPos((int)x, (int)y);
         QWindow *tlw = topLevelWindowAt(globalPos);
-        QPointF localPos = tlw ? (globalPosF - tlw->position()) : globalPosF;
+
+        QPointF localPos = globalPosF;
+        if (tlw) {
+            QPlatformWindow *platformWindow = tlw->handle();
+            localPos = platformWindow ? platformWindow->mapFromGlobal(globalPos) : globalPosF;
+        }
 
         // Galaxy Note with plain Android:
         // 0 1 0    stylus press
-- 
2.20.1.windows.1

